File: src/com/madhu/scala/tdop/TDOP.scala

package com.madhu.scala.tdop

import java.util.regex.Pattern
import java.util.regex.Matcher
import scala.collection.mutable.ArrayBuffer
import scala.collection.mutable.Stack

sealed class TerminalID(val name: String, val regex: String) {
  val pattern: Pattern = Pattern.compile(regex)
  var matcher: Matcher = null

  def init(input: String) = matcher = pattern.matcher(input)
  override def toString(): String = name
}

class Terminal(val name: String, value: String) {
  override def toString(): String = String.format("%s", value)
  def matches(that: TerminalID) = this.name.equals(that.name)
  def matches(that: String) = this.name.equals(that)
}

class SyntaxError(val message: String) extends RuntimeException(message) {
}

class Scanner(
    val input: String,
    val patterns: List[TerminalID],
    var lookAhead: Option[Terminal] = None,
    var terminal: Terminal = null,
    var column: Int = 0) {
  def this(input: String) = {
    this(input, List(
      new TerminalID("(", "\\("),
      new TerminalID(")", "\\)"),
      new TerminalID("&&", "\\&\\&"),
      new TerminalID("||", "\\|\\|"),
      new TerminalID("!=", "\\!\\="),
      new TerminalID("==", "\\=\\="),
      new TerminalID("+", "\\+"),
      new TerminalID("-", "\\-"),
      new TerminalID("*", "\\*"),
      new TerminalID("/", "\\/"),
      new TerminalID("INT", "[0-9]+")
    ))
    patterns.foreach(p => p.init(input))
    lookAhead = next
  }

  def next(): Option[Terminal] = {
    while (column < input.length() && input.charAt(column).isWhitespace) {
      column += 1
    }
    if (column >= input.length()) {
      return None
    }
    patterns.foreach(p => {
      val m = p.matcher
      if (m.find(column) && m.start() == column) {
        val end = m.end()
        val t = new Terminal(p.name, input.substring(column, end))
        column = end
        return Some(t)
      }
    })
    throw new SyntaxError("Unrecognized input: " + input.charAt(column))
  }

  def matches(types: String*): Boolean = {
    if (lookAhead == None) {
      return false
    }
    types.foreach(t => {
      lookAhead match {
        case Some(l) => {
          if (l.matches(t)) {
            terminal = l
            lookAhead = next
            return true
          }
        }
        case None =>
      }
    })
    false
  }

  def expect(types: String*): Terminal = {
    if (matches(types:_*)) {
      return terminal
    }
    throw new SyntaxError("Expected one of " + types.mkString(" ") + ", found " + lookAhead.getOrElse("?null"))
  }

  def atEnd(): Boolean = {
      lookAhead match {
        case Some(l) => false
        case None => true
      }
  }
}

class Tree[V](val value: V, val children: ArrayBuffer[Tree[V]] = new ArrayBuffer[Tree[V]]()) {
  def iterator = children.iterator

  def apply(index: Int): Tree[V] = {
    children(index)
  }

  def +=(that: Tree[V]): Tree[V] = {
    children += that
    this
  }

  def +=(that: V): Tree[V] = {
    children += new Tree(that)
    this
  }

  def size(): Int = children.size

  def isLeaf(): Boolean = size == 0

  override def toString(): String = {
    val sb: StringBuilder = new StringBuilder()
    doToString(sb)
    sb.toString
  }

  def doToString(sb: StringBuilder) {
    if (isLeaf) {
      sb ++= value.toString
    } else {
      sb += '('
      sb ++= value.toString
      children.foreach(c => {
        sb += ' '
        c.doToString(sb)
      })
      sb += ')'
    }
  }
}

class Parser(val sc: Scanner,
    var prec: Array[Array[String]] = null,
    var stack: Stack[Tree[Terminal]] = new Stack()) {
  def this(input: String) = {
    this(new Scanner(input))
    prec = Array(Array("&&", "||"), Array("==", "!="), Array("+", "-"), Array("*", "/"))
  }

  def parse(): Tree[Terminal] = {
    val tree = parseExp
    if (!sc.atEnd) {
      throw new SyntaxError("Unexpected input " + sc.terminal)
    }
    tree
  }

  def parseExp(): Tree[Terminal] = {
    parseX(0)
  }

  def parseX(index: Int): Tree[Terminal] = {
    var result = parseY(index)
    while (sc.matches(prec(index) : _*)) {
      val temp = new Tree(sc.terminal)
      temp += result
      temp += parseY(index)
      result = temp
    }
    result
  }

  def parseY(index: Int): Tree[Terminal] = {
    if (index >= prec.length-1) return parsePrim
    parseX(index + 1)
  }

  def parsePrim(): Tree[Terminal] = {
    if (sc.matches("(")) {
      val tree = parseExp
      sc.expect(")")
      return tree
    }
    if (sc.matches("-")) {
      val tree = new Tree(sc.terminal)
      tree += parsePrim
      return tree
    }
    new Tree(sc.expect("INT"))
  }
}

object ParserTest {
  def main(args: Array[String]): Unit = {
    val p = new Parser("1+2* 4 == 5/6 && 7 != 8")
    println(p.parse)
  }
}

File: py/TDOP.py

import re

class Pattern(object):
    def __init__(self, name, regex):
        self.name = name
        self.pattern = re.compile(regex)
        
    def match(self, input, index):
        return self.pattern.match(input, index)
    
class Terminal(object):
    def __init__(self, name, value):
        self.name = name
        self.value = value

    def __str__(self):
        if self.name.lower() == self.value:
            return self.name
        return '%s(%s)' % (self.name, self.value)
        
class Tree(object):
    def __init__(self, *values):
        self.value = None
        if len(values) > 0:
            self.value = values[0]
        if len(values) > 1:
            self.children = [x for x in values[1:]]
        else:
            self.children = []

    def add(self, value):
        if isinstance(value, Tree):
            self.children.append(value)
        else:
            self.children.append(Tree(value))
        return self
        
    def __len__(self):
        return len(self.children)

    def isLeaf(self):
        return len(self.children) == 0

    def toDot(self):
        self.label(1)
        return 'digraph ast {%s\n}' % self.doToDot('')

    def label(self, id):
        self.id = id
        id += 1
        for c in self.children:
            id = c.label(id)
        return id

    def doToDot(self, dot):
        dot = '%s\n%d [label="%s"];' % (dot, self.id, self.value.value)
        for c in self.children:
            dot = c.doToDot(dot)
            dot = '%s\n%d -> %d;' % (dot, self.id, c.id)
        return dot

    def __str__(self):
        if self.isLeaf():
            return self.value.__str__()
        result = '(%s)' % self.value
        for c in self.children:
            result = '%s %s' % (result, c)
        return '%s)' % result

class Scanner(object):
    def __init__(self, input, patterns):
        self.input = input
        self.index = 0
        self.patterns = patterns
        self.terminal = None
        self.lookAhead = self.next()

    def next(self):
        while self.index < len(self.input) and self.input[self.index].isspace():
            self.index += 1
        if self.index >= len(self.input):
            return None
        for p in self.patterns:
            match = p.match(self.input, self.index)
            if match:
                self.index = match.end()
                return Terminal(p.name, match.group())
        raise Exception('Unrecognized input: %s' % (self.input[self.index]))
        
    def matches(self, *types):
        if self.lookAhead == None:
            return False
        for t in types:
            if t == self.lookAhead.name:
                self.terminal = self.lookAhead
                self.lookAhead = self.next()
                return True
        return False

    def expect(self, *types):
        if self.matches(*types):
            return self.terminal
        raise Exception('Expected %s, found %s' % (','.join(types), self.lookAhead))

    def atEnd(self):
        return self.lookAhead == None

class Parser(object):
    def __init__(self, scanner):
        self.sc = scanner
        self.prec = [('&&','||'), ('==','!=','>','<','>=','<='), ('+','-'), ('*','/','%')]

    def parse(self):
        tree = self.parseStatement()
        if not self.sc.atEnd():
            raise Exception('Unexpected input: %s' % self.sc.terminal)
        return tree

    def parseStatement(self):
        if self.sc.matches('{'):
            tree = Tree(self.sc.terminal)
            while not self.sc.matches('}'):
                tree.add(self.parseStatement())
            return tree
        if self.sc.matches('WHILE'):
            return Tree(self.sc.terminal, self.parseExp(), self.parseStatement())
        if self.sc.matches('BREAK'):
            tree = Tree(self.sc.terminal)
            self.sc.expect(';')
            return tree
        if self.sc.matches('IF'):
            tree = Tree(self.sc.terminal, self.parseExp(), self.parseStatement())
            if self.sc.matches('ELSE'):
                tree.add(self.parseStatement())
            return tree
        if self.sc.matches('ID'):
            id = self.sc.terminal
            if self.sc.matches('='):
                tree = Tree(self.sc.terminal, Tree(id), self.parseExp())
                self.sc.expect(';')
                return tree
        self.sc.expect(';')

    def parseExp(self):
        return self.parseHead(0)

    def parseHead(self, index):
        result = self.parseTail(index)
        while self.sc.matches(*self.prec[index]):
            result = Tree(self.sc.terminal, result, self.parseTail(index))
        return result
        
    def parseTail(self, index):
        if index >= len(self.prec)-1:
            return self.parsePrim()
        return self.parseHead(index + 1)

    def parsePrim(self):
        if self.sc.matches('('):
            tree = self.parseExp()
            self.sc.expect(')')
            return tree
        if self.sc.matches('-'):
            return Tree(self.sc.terminal, self.parsePrim())
        return Tree(self.sc.expect('INT', 'ID'))
                    
if __name__ == '__main__':
    patterns = []
    patterns.append(Pattern('INT', r'[0-9]+'))
    patterns.append(Pattern('IF', r'if'))
    patterns.append(Pattern('ELSE', r'else'))
    patterns.append(Pattern('WHILE', r'while'))
    patterns.append(Pattern('BREAK', r'break'))
    patterns.append(Pattern('ID', r'[a-zA-Z][a-zA-Z0-9_]*'))
    patterns.append(Pattern(';', r'\;'))
    patterns.append(Pattern('{', r'\{'))
    patterns.append(Pattern('}', r'\}'))
    patterns.append(Pattern('[', r'\['))
    patterns.append(Pattern(']', r'\]'))
    patterns.append(Pattern('(', r'\('))
    patterns.append(Pattern(')', r'\)'))
    patterns.append(Pattern('+', r'\+'))
    patterns.append(Pattern('-', r'\-'))
    patterns.append(Pattern('*', r'\*'))
    patterns.append(Pattern('/', r'\/'))
    patterns.append(Pattern('<=', r'\<\='))
    patterns.append(Pattern('>=', r'\>\='))
    patterns.append(Pattern('==', r'\=\='))
    patterns.append(Pattern('!=', r'\!\='))
    patterns.append(Pattern('&&', r'\&\&'))
    patterns.append(Pattern('||', r'\|\|'))
    patterns.append(Pattern('=', r'\='))
    patterns.append(Pattern('<', r'\<'))
    patterns.append(Pattern('>', r'\>'))
    patterns.append(Pattern('%', r'\%'))

    input = '''
    {
        i = 0;
        while i<10 {
            a = 2*3;
            if i % 1 == 0 {
                a = a + 1;
            } else {
                a = a + 2;
            }
            i = i+1;
        }
    }
    '''
    
    p = Parser(Scanner(input, patterns))
    dot = p.parse().toDot()
    print(dot)
    